ogre2/src/Ogre2SegmentationCamera.cc

```diff
@@ -0,0 +1,858 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+
+#include <random>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "ignition/common/Console.hh"
+#include "ignition/math/Color.hh"
+#include "ignition/rendering/ogre2/Ogre2Camera.hh"
+#include "ignition/rendering/ogre2/Ogre2Conversions.hh"
+#include "ignition/rendering/ogre2/Ogre2Includes.hh"
+#include "ignition/rendering/ogre2/Ogre2RenderTarget.hh"
+#include "ignition/rendering/ogre2/Ogre2RenderTypes.hh"
+#include "ignition/rendering/ogre2/Ogre2RenderEngine.hh"
+#include "ignition/rendering/ogre2/Ogre2Scene.hh"
+#include "ignition/rendering/ogre2/Ogre2SegmentationCamera.hh"
+#include "ignition/rendering/ogre2/Ogre2SelectionBuffer.hh"
+#include "ignition/rendering/ogre2/Ogre2Visual.hh"
+#include "ignition/rendering/RenderTypes.hh"
+#include "ignition/rendering/Utils.hh"
+
+namespace ignition
+{
+namespace rendering
+{
+inline namespace IGNITION_RENDERING_VERSION_NAMESPACE {
+/// \brief Helper class to assign unique colors to renderables
+class Ogre2SegmentationMaterialSwitcher : public Ogre::RenderTargetListener
+{
+  /// \brief Constructor
+  /// \param[in] _scene Ogre scene to be rendered
+  public: explicit Ogre2SegmentationMaterialSwitcher(Ogre2ScenePtr _scene);
+
+  /// \brief Destructor
+  public: ~Ogre2SegmentationMaterialSwitcher() = default;
+
+  /// \brief Ogre's pre render update callback
+  /// \param[in] _evt Ogre render target event containing information about
+  /// the source render target.
+  public: virtual void preRenderTargetUpdate(
+              const Ogre::RenderTargetEvent &_evt) override;
+
+  /// \brief Ogre's post render update callback
+  /// \param[in] _evt Ogre render target event containing information about
+  /// the source render target.
+  public: virtual void postRenderTargetUpdate(
+              const Ogre::RenderTargetEvent &_evt) override;
+
+  /// \brief Convert label of semantic map to a unique color for colored map
+  /// \param[in] _label id of the semantic map or encoded id of panoptic map
+  /// \param[in] _isMultiLink bool used to skip the taken color check if the
+  /// label is for a multi link model, as all links should have the same color
+  /// \return _color unique color in the colored map for that label
+  private: math::Color LabelToColor(int64_t _label,
+    bool _isMultiLink = false);
+
+  /// \brief Check if the color is taken previously
```

Comment:
can you add a small comment that it also adds the color to the list if it's not taken?